### ISSUE TITLE
Installation on Trusty Fails with "libfastjson.so.3: cannot open shared object file"

### ISSUE THREAD

**Issue by felixhummel** [rsyslog/rsyslog-pkg-ubuntu#46]
Steps taken:
- follow instructions from http://www.rsyslog.com/ubuntu-repository/

Expected:
- it works

Actual:
```
$ rsyslogd -v
rsyslogd: error while loading shared libraries: libfastjson.so.3: cannot open shared object file: No such file or directory
```

More info:
```
$ wajig show libfastjson
Package: libfastjson                     
New: yes
State: installed
Automatically installed: yes
Version: 0.99.3-0adiscon1trusty1
Priority: extra
Section: libs
Maintainer: Florian Riedl <friedl@adiscon.com>
Architecture: amd64
Uncompressed Size: 69,6 k
Depends: libc6 (>= 2.14)
Conflicts: libfastjson
Description: LibFastJSON
 The 'libfastjson' library contains the a modified json-c library with better performance 
 
 This package contains the shared library.

$ wajig list-files libfastjson
/.
/usr
/usr/lib
/usr/lib/libfastjson.so.4.0.0
/usr/share
/usr/share/doc
/usr/share/doc/libfastjson
/usr/share/doc/libfastjson/copyright
/usr/share/doc/libfastjson/changelog.Debian.gz
/usr/lib/libfastjson.so.4

$ cat /etc/lsb-release 
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=14.04
DISTRIB_CODENAME=trusty
DISTRIB_DESCRIPTION="Ubuntu 14.04.4 LTS"
```

**Comment by suan** [comment]
Can't spin up new instances as of today due to this bug as well, any recommended workarounds at least?

**Comment by felixhummel** [comment]
I downgraded.

**Comment by suan** [comment]
@felixhummel Was there a repo you could get the so.3 version from? Or did you manually install from a `.deb`?

**Comment by felixhummel** [comment]
@suan I downgraded rsyslog to 7.4.4 (the default on trusty).

**Comment by davidelang** [comment]
This is due to the new libfastjson that was released today. If you downgrade it 
to 0.99.2 everything should work with the latest rsyslog.

the problem is that rsyslog, liblognorm, and mmnormalize need to be compiled 
against the new version of libfastjson.

The .so version of the library changed because there are changes in it that are 
not backwards compatible.

When the new rsyslog release hits in the next few days, everything will work 
together again.

In the meantime, either downgrade to the older libfastjson, or compile 8.19 
against the new version.


It looks like the release procedure needs to be changed so that when a library 
is updated, the dependent packages get rebuilt (8.19.1 or something)

David Lang

**Comment by suan** [comment]
Thanks, David.

> It looks like the release procedure needs to be changed so that when a library
is updated, the dependent packages get rebuilt (8.19.1 or something)

Either that, or keep the old versions available in the repo during transitions, so apt won't blow up?

**Comment by davidelang** [comment]
On Mon, 11 Jul 2016, Suan Yeo wrote:

>> It looks like the release procedure needs to be changed so that when a library
> is updated, the dependent packages get rebuilt (8.19.1 or something)
>
> Either that, or keep the old versions available in the repo during transitions, so apt won't blow up?

The problem is that the dependencies are on version >= X, so when a new version 
comes out, it will be pulled automatically.

I'm pretty sure the repo contains all versions that have been released, so the 
old 0.99.2 version should still be there.

David Lang

**Comment by rgerhards** [comment]
2016-07-12 2:45 GMT+02:00 davidelang <notifications@github.com>:

> On Mon, 11 Jul 2016, Suan Yeo wrote:
>
> >> It looks like the release procedure needs to be changed so that when a
> library
> > is updated, the dependent packages get rebuilt (8.19.1 or something)
> >
> > Either that, or keep the old versions available in the repo during
> transitions, so apt won't blow up?
>
> The problem is that the dependencies are on version >= X, so when a new
> version
> comes out, it will be pulled automatically.
>
> I'm pretty sure the repo contains all versions that have been released, so
> the
> old 0.99.2 version should still be there.
>

I am far from being an expert at this, but I  think this actually is a
launchpad issue. From what I read here, they seem to remove previous
versions when a new one is released and keep only the most recent one. That
of course breaks many things, like we see here.

In the past, it was never a problem to release an updated version, because
the versioning scheme makes sure that the old version is being used. It's
simply impossible in general to build all depending packages freshly if
some library changes. In this context, of course, it is simpler, but that
doesn't apply to the general principle. So I wonder if we have set
something wrong in launchpad ... or if launchpad is really not up to
provide a reliable repository.

If someone knows if there is a setting, or anything else, that improves the
situation advise is very welcome. I also tried to rebuild rsyslog inside
launchpad (the current version), but I don't see any ability at all to do
that. I am even less an expert on launchpad, so I may not have searched the
right places. I'll check with Florian as he comes in. Also, any advise on
how to make launchpad rebuild the rsyslog package with the new library
versions would be appreciated.

Rainer
